### PR TITLE
add flake configuration (devshell + package)

### PR DESCRIPTION
**Pull request by Sigmanificient**
I, I love your project! I'd like to ship it for my linux distribution, [NixOS](https://nixos.org/).

I added a `flake.nix` & `flake.lock`  file, to provide a way to build a [nix derivation](https://nixos.org/manual/nix/stable/language/derivations.html) (package) for it.

Build `fzg-make` using the nix flake:
```sh
nix build .
```

> Note: [Flakes](https://nixos.wiki/wiki/Flakes) are still an "experimental feature" of nix despite its wide use.

Moreover, this `flake.nix` provides a [shell](https://nixos.org/manual/nix/stable/command-ref/nix-shell.html) with the dependencies required to compile this project.
This means that in any Linux distributions, windows WSL or even OSX, and other system supporting the [nix package manager](https://nixos.org/download) you use use it to pull the dependencies and enters in a development shell with everything setup for you.

Entering the shell
```sh
nix develop
# do casual stuff
make
```

This can integrate with [direnv](https://direnv.net/) to automatically loads thoses dependencies went you enter the directory, once enabled.
direnv can integrate within multiples shells (bash, zsh, fish, tcsh, ...)

How to setup automatic direnv in project:
```sh
echo "use flake" | tee -a .envrc
direnv allow
```

### CODE REVIEW

**Comment by kyu08** [comment]
> I, I love your project! I'd like to ship it for my linux distribution, [NixOS](https://nixos.org/).

Thank you so much! I'm glad to hear that.
And I appreciate you to create this PR.

I have one question. (I'm not familiar with Linux distributions.)

I found this PR ([fzf-make: init at 0.6.0 #250014](https://github.com/NixOS/nixpkgs/pull/250014)).

I am wondering if the above PR might allow fzf-make can be used on nixOS without having to merge this PR you created, is that correct? (Also, it seems a bit odd to add code for a specific package manager to this repository.)

**Comment by Sigmanificient** [comment]
The pull request your repository would allow for 3 things:
- `nix develop`: Create "dev shell", that have the dependencies to work on the projects
-  `nix build`: Building the project locally, from the flake (not bound to a specific version)
- `nix run github:kyu08/fzf-make` to build and run from the repository (will download it automatically)

The repo you mentioned is the official repository for nix to fetch its packages. 
Having your package merged into it, would mean that people could download your application globally as a part of they're nix config, with the version explicitly defined. It would also be put into the package search website for anyone to see.

**Comment by kyu08** [comment]
Ah! So you mean this PR is to set up a development environment using nix! (Sorry it took me so long to understand.)

Then could you please show the development flow in `README.md` following below format?

```md
# 📦 Installation
...
# Develop
## nix

<!-- Please write here! -->

# 💡 Usage
```

And if you don't mind, it would be very helpful if you could add to the `README.md` how to install fzf-make on nixOS. 🙇

File: flake.nix [hunk 1]
```diff
@@ -0,0 +1,35 @@
+{
+  description = "Fuzzy finder for Makefile";
```
Comment (kyu08) [inline flake.nix:2]: I'm sorry. I rewrote description of this repository a while ago 😧
So please change the description field as follows.

```suggestion
  description = "fzf-make is the command line tool that execute make target using fuzzy finder with preview window.";
```

File: flake.nix [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+            bat
+            cargo
+            gnumake
+          ];
+        };
+
+        packages = rec {
+          fzf-make = pkgs.rustPlatform.buildRustPackage {
+            pname = "fzf-make";
+            version = "0.6.0";
```
Comment (kyu08) [inline flake.nix:25]: Sorry about this too.
I released `v0.7.0` after you created this PR. 🙇 
So please change as follows.

```suggestion
            version = "0.7.0";
```

**Review by kyu08 — COMMENTED** [review]
Please make only two minor changes. 🙏

**Comment by kyu08** [comment]
@Sigmanificient 
Could you resolve conflicts please? 🙇‍♂️(My changes after you created this PR caused this, sorry.)
Also, I would like you to add installation instructions for fixOS in the `README.md`.

File: README.md [hunk 3]
```diff
@@ -35,15 +35,35 @@ brew update
 brew upgrade fzf-make
 ```
 
-## Arch Linux
+### nix
```
Comment (kyu08) [inline README.md:38]: Here is the installation section.
I want you to write how to develop using nix [here](https://github.com/kyu08/fzf-make/blob/0217175558ac105e5cf048085354c34d14eded6d/README.md#L55)(between `# 💡 Usage` and `# 👥 Contribution`) starting by `# 💻 Development`.

Apart from that, I would like you to describe how to install fzf-make on nixOS in the installation section in the following format.

```md
# 📦 Installation
...

## Arch Linux
...

## NixOS
<!-- please write how to install to NixOS here! -->
...

## OS-independent method
...
```

If you have any questions, feel free to ask me. 😄

File: README.md [hunk 4]
```diff
@@ -35,15 +35,35 @@ brew update
...
+```sh
+nix build .
+```
+
+> Or run it from your terminal
+
+```sh
+nix run github:kyu08/fzf-make
+
+### Arch Linux
```
Comment (kyu08) [inline README.md:57]: Please fix this like below!

```suggestion
## Arch Linux
```

File: README.md [hunk 5]
```diff
@@ -35,15 +35,35 @@ brew update
...
 
 `fzf-make` can be installed from the [AUR](https://aur.archlinux.org/packages/fzf-make) using an [AUR helper](https://wiki.archlinux.org/title/AUR_helpers). For example:
 
 ```sh
 paru -S fzf-make
 ```
 
-## OS-independent method
+### OS-independent method
+
```
Comment (kyu08) [inline README.md:66]: Please fix this like below and remove empty new line!

```suggestion
## OS-independent method
```

**Review by kyu08 — COMMENTED** [review]
Thanks for responding!
There are a few things I would like to see fixed. 😀

File: README.md [hunk 6]
```diff
@@ -43,7 +43,24 @@ brew upgrade fzf-make
 paru -S fzf-make
 ```
 
+## NixOS / Nix (package manager)
+
```
Comment (kyu08) [inline README.md:47]: [NITS]
Please remove this empty line.

File: README.md [hunk 7]
```diff
@@ -43,7 +43,24 @@ brew upgrade fzf-make
...
+nix run github:kyu08/fzf-make
+```
+
+Or from the nixpkgs (channel >= 23.05)
+```sh
+nix run nixpkgs#fzf-make
+```
+
+> **Note**
+> You may need to enable experimental feature. In that case, write the following command to enable them
```
Comment (kyu08) [inline README.md:59]: [NITS]
I think this is better.

```suggestion
> You may need to enable experimental feature. In that case, execute the following command to enable them
```

File: README.md [hunk 8]
```diff
@@ -43,7 +43,24 @@ brew upgrade fzf-make
...
+
+`fzf-make` can be run from the repository (latest version)
+```sh
+nix run github:kyu08/fzf-make
+```
+
+Or from the nixpkgs (channel >= 23.05)
+```sh
+nix run nixpkgs#fzf-make
+```
```
Comment (kyu08) [inline README.md:56]: This command only executes fzf-make once, and does not appear to be an install command.
I would like to see the install procedure written here. (`nix-env -iA nixpkgs.${package-name}` or `nix profile install nixpkgs#${package-name}` seem applicable)

File: README.md [hunk 9]
```diff
@@ -43,7 +43,24 @@ brew upgrade fzf-make
...
+
+> **Note**
+> You may need to enable experimental feature. In that case, write the following command to enable them
+> `echo "experimental-features = nix-command flakes" | tee  ~/.config/nix/nix.conf`
+
 ## OS-independent method
+
```
Comment (kyu08) [inline README.md:63]: [NITS]
Please remove this empty line.

File: README.md (hunk 8 above)
Comment (Sigmanificient) [reply README.md:56]: > This command only executes fzf-make once, and does not appear to be an install command.
I would like to see the install procedure written here. 

Yes, but `nix-env -iA` is depreciated, the proper to have it install is to edit the configuration.nix file and either add it to the user packages, or system ones. There is also the possibility to add it on user packages with home manager, or as a flake input using the url, ... and there is many other ways, which is why I'm not sure how to approach it.

I'm not really aware of the `nix profile` command too.

Also, it run the command once, but it put the flake evaluated derivation within the cache, which allows you to run it multiple times without rebuilding it. That the way i use it tbh, it's way easier.

**Comment by Sigmanificient** [comment]
I fixed both line & type, waiting for your input about installation method

**Review by kyu08 — APPROVED** [review]
@Sigmanificient 
Thank you for taking the time to explain it to me as I am unfamiliar with NixOS.
I now understand that there are several ways of installing an application. If I think about it, it seems that if one example is shown, users can read it into their preferred method based on that. (I may have overthought it a bit.  Your input helped me make the right decision.)

Thanks for creating this PR. I look forward to your contributions in the future! Thanks!

**Comment by Sigmanificient** [comment]
This is one of the great strength of nix, but also can be problematic for new users. Everything is interconnected in many many ways

**Comment by kyu08** [comment]
@Sigmanificient 
Hi, thank you for your helping me to merge this PR the other day!
I have one question about nix.
When I release next version, what I should do in connection with nix is just update [version in `flake.nix`](https://github.com/kyu08/fzf-make/blob/d57079fb7b1deb3c5d88f597f8ac9d337791d33a/flake.nix#L25)?

**Comment by Sigmanificient** [comment]
Update the string version and maybe run `nix flake update` in case some inputs changed I believe.
If you add more dependencies, I could take a look at it to update the flake accordingly.

**Comment by kyu08** [comment]
Thanks for your input!

**Pull request merged**